### PR TITLE
fix ingester: grpc method=/logproto.Querier/Query fail.

### LINKED ISSUE #5964: ingester: grpc method=/logproto.Querier/Query fail.
**Describe the bug**
rpc error: code = Internal desc = transport: transport: the stream is done or WriteHeader was already called.

image

image

**To Reproduce**
1. Started Loki (SHA or version)
2. Started Promtail (SHA or version) to tail '...'
3. Query: `{} term`

**Expected behavior**

**Environment:**
 - Infrastructure: [e.g., Kubernetes, bare-metal, laptop]
 - Deployment tool: [e.g., helm, jsonnet]

**Screenshots, Promtail config, or terminal output**
If applicable, add any output to help explain your problem.

### PR DESCRIPTION
**Pull request by liguozhong**
**What this PR does / why we need it**:
When ingester's ctx times out, don't try to use grpc to return data to the querier again. 
This PR fixes the ingester error rate > 0 for our loki cluster.
We rely on this error rate indicator for alarming
image

**Which issue(s) this PR fixes**:
Fixes #5964

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [ ] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Review by kavirajk — COMMENTED** [review]
changes looks reasonable. Would wait for someone more familiar with ingester to do final check. 
Just to double check this doesn't break any existing behaviour (e.g: querier may still need stats about what happened)

**Comment by bboreham** [comment]
It's not something you added in this PR, but I wanted to note there is a simpler way to do `isDone()`: https://github.com/grafana/loki/blob/557e6e528059691911945bd7a26e8efded84cd08/pkg/ingester/instance.go#L683-L690

Just check `ctx.Err() != nil`, as in the [docs](https://pkg.go.dev/context#Context):
```
	// If Done is not yet closed, Err returns nil.
	// If Done is closed, Err returns a non-nil error explaining why:
```

**Comment by bboreham** [comment]
I find it slightly weird that there are now four calls to `isDone()` in `sendBatches()`, three looking at `ctx` and one looking at `queryServer.Context()`.

If there is a good reason for this, please put it in a comment.

Possibly both loops should check both contexts.
Arguably both loops should only check at one point.

File: pkg/ingester/instance.go [hunk 1]
```diff
@@ -734,7 +736,9 @@ func sendBatches(ctx context.Context, i iter.EntryIterator, queryServer QuerierQ
 
 		stats.AddIngesterBatch(int64(batchSize))
 		batch.Stats = stats.Ingester()
-
+		if isDone(ctx) {
+			break
+		}
 		if err := queryServer.Send(batch); err != nil {
 			return err
```
Comment (bboreham) [inline pkg/ingester/instance.go:743]: If the intention is that context cancellation should never result in an error return from this function then I would think you'd want to swallow that here too. (and in similar places)

**Comment by liguozhong** [comment]
> It's not something you added in this PR, but I wanted to note there is a simpler way to do `isDone()`:
> 
> https://github.com/grafana/loki/blob/557e6e528059691911945bd7a26e8efded84cd08/pkg/ingester/instance.go#L683-L690
> 
> Just check `ctx.Err() != nil`, as in the [docs](https://pkg.go.dev/context#Context):
> 
> ```
> 	// If Done is not yet closed, Err returns nil.
> 	// If Done is closed, Err returns a non-nil error explaining why:
> ```

thanks , done.

**Comment by liguozhong** [comment]
> I find it slightly weird that there are now four calls to `isDone()` in `sendBatches()`, three looking at `ctx` and one looking at `queryServer.Context()`.
> 
> If there is a good reason for this, please put it in a comment.
> 
> Possibly both loops should check both contexts. Arguably both loops should only check at one point.

	_, ctx := stats.NewContext(queryServer.Context())

I checked that ctx comes from queryServer.Context() , so I deleted queryServer.Context() and uniformly used ctx as the parameter of isDone() func.

**Comment by splitice** [comment]
I like this commit. These messages have similarly annoyed us during monitoring.

**Review by bboreham — APPROVED** [review]
Thanks for responding to my comments.  LGTM.

**Comment by DylanGuedes** [comment]
Do you mind rebasing it? Other than that, LGTM

**Review by DylanGuedes — APPROVED** [review]

**Comment by splitice** [comment]
Go away useless bot

**Comment by MichelHollands** [comment]
@liguozhong Can you rebase? Once that is done this can be merged.

**Comment by liguozhong** [comment]
> @liguozhong Can you rebase? Once that is done this can be merged.

done,thanks

**Pull request merged**